src/ext4.c

```diff
@@ -2016,23 +2016,25 @@ int ext4_fwrite(ext4_file *file, const void *buf, size_t size, size_t *wcnt)
 	return r;
 }
 
-int ext4_fseek(ext4_file *file, uint64_t offset, uint32_t origin)
+int ext4_fseek(ext4_file *file, int64_t offset, uint32_t origin)
 {
 	switch (origin) {
 	case SEEK_SET:
-		if (offset > file->fsize)
+		if (offset < 0 || (uint64_t)offset > file->fsize)
 			return EINVAL;
 
 		file->fpos = offset;
 		return EOK;
 	case SEEK_CUR:
-		if ((offset + file->fpos) > file->fsize)
+		if ((offset < 0 && (uint64_t)(-offset) > file->fpos) ||
+		    (offset > 0 &&
+		     (uint64_t)offset > (file->fsize - file->fpos)))
 			return EINVAL;
 
 		file->fpos += offset;
```

Comment:
It is equivalent to ```file->fpos = file->fpos + offset;```. I think we need to check if there is not implicit conversion ```offset``` variable to uint64_t

Reply:
That's right. Let me work on the test cases first. Then I will fix this.

Reply:
It sholud work without casting if we assume complement two encoding of signed values.

Reply:
Hmm, in Linux ext4 driver file->fpos is signed and makes everything much easier. I think that might be easier and more error-prone to also make fpos signed here. Mixing unsigned and signed is almost always a bad idea...

Reply:
Changing file->fpos (and maybe file->fsize) to int64 seems to be a good idea. IMO we should do it as a separate PR.